### DISCUSSION TITLE
AttributeError: module 'cupy' has no attribute 'cublas'

### DISCUSSION THREAD

**Discussion by jayshow03** [explosion/spaCy#10277] [Help: Installation] [answered]
Hi there, does anyone know the answer to this question?  I tried different versions of cupy-cuda110 but it did not work so far.  Below you can see the unexpected information. Hope someone can address it, thanks!
(pytorch) C:\Users\Jayshow\PycharmProjects\spacy_transform>pip list
WARNING: Ignoring invalid distribution -upy-cuda110 (c:\users\jayshow\anaconda3\envs\pytorch\lib\site-packages)
Package                  Version
------------------------ ---------
aiofiles                 0.8.0
aiohttp                  3.8.1
aiosignal                1.2.0
altair                   4.2.0
argon2-cffi              20.1.0
async-generator          1.10
async-timeout            4.0.1
attrs                    21.4.0
backcall                 0.2.0
bleach                   4.1.0
blis                     0.7.5
boto3                    1.20.24
botocore                 1.23.24
brotlipy                 0.7.0
cachetools               5.0.0
catalogue                2.0.6
certifi                  2021.10.8
cffi                     1.15.0
charset-normalizer       2.0.11
click                    7.1.2
colorama                 0.4.4
cryptography             36.0.0
cupy-cuda110             8.0.0
cymem                    2.0.6
debugpy                  1.5.1
decorator                5.1.1
defusedxml               0.7.1
Deprecated               1.2.13
docker                   5.0.3
entrypoints              0.3
et-xmlfile               1.1.0
fastapi                  0.68.2
fastrlock                0.5
filelock                 3.4.2
frozenlist               1.2.0
gobbli                   0.2.4
google-api-core          1.25.1
google-auth              1.33.0
google-cloud-core        1.7.1
google-cloud-storage     1.41.0
google-crc32c            1.1.2
google-resumable-media   1.3.1
googleapis-common-protos 1.53.0
grpcio                   1.42.0
h11                      0.13.0
huggingface-hub          0.0.8
humanize                 3.14.0
idna                     3.3
importlib-metadata       4.8.2
ipykernel                6.4.1
ipython                  7.31.1
ipython-genutils         0.2.0
ipywidgets               7.6.5
jedi                     0.18.1
Jinja2                   3.0.3
jmespath                 0.10.0
joblib                   1.1.0
jsonschema               3.2.0
jupyter                  1.0.0
jupyter-client           7.1.2
jupyter-console          6.4.0
jupyter-core             4.9.1
jupyterlab-pygments      0.1.2
jupyterlab-widgets       1.0.0
langcodes                3.3.0
MarkupSafe               2.0.1
matplotlib-inline        0.1.2
mistune                  0.8.4
mkl-fft                  1.3.1
mkl-random               1.2.2
mkl-service              2.4.0
msgpack                  1.0.3
multidict                5.1.0
murmurhash               1.0.6
nbclient                 0.5.3
nbconvert                6.1.0
nbformat                 5.1.3
nest-asyncio             1.5.1
numpy                    1.22.2
olefile                  0.46
openpyxl                 3.0.9
packaging                21.3
pandas                   1.4.0
pandocfilters            1.5.0
parso                    0.8.3
pathy                    0.6.1
peewee                   3.14.8
pickleshare              0.7.5
Pillow                   8.2.0
pip                      21.2.2
plac                     1.1.3
preshed                  3.0.6
prodigy                  1.11.6
prometheus-client        0.13.1
prompt-toolkit           3.0.20
protobuf                 3.19.1
pyasn1                   0.4.8
pyasn1-modules           0.2.8
pycparser                2.21
pydantic                 1.8.2
Pygments                 2.11.2
PyJWT                    2.3.0
pyOpenSSL                22.0.0
pyparsing                3.0.7
pyrsistent               0.18.0
PySocks                  1.7.1
python-dateutil          2.8.2
pytz                     2021.3
pywin32                  227
pywinpty                 2.0.2
PyYAML                   6.0
pyzmq                    22.3.0
qtconsole                5.2.2
QtPy                     1.11.2
ray                      1.10.0
redis                    4.1.3
regex                    2021.8.3
requests                 2.27.1
rsa                      4.7.2
s3transfer               0.5.0
sacremoses               0.0.43
scikit-learn             1.0.2
scipy                    1.8.0
Send2Trash               1.8.0
setuptools               58.0.4
shellingham              1.3.1
sip                      4.19.13
six                      1.16.0
smart-open               5.2.1
spacy                    3.2.0
spacy-alignments         0.8.4
spacy-legacy             3.0.8
spacy-loggers            1.0.1
spacy-transformers       1.1.4
srsly                    2.4.2
starlette                0.14.2
terminado                0.13.1
testpath                 0.5.0
thinc                    8.0.13
threadpoolctl            3.1.0
tokenizers               0.10.3
toolz                    0.11.2
torch                    1.7.1
torchvision              0.8.2
tornado                  6.1
tqdm                     4.62.3
traitlets                5.1.1
transformers             4.6.0
typer                    0.3.2
typing-extensions        3.10.0.2
urllib3                  1.26.8
uvicorn                  0.13.4
wasabi                   0.9.0
wcwidth                  0.2.5
webencodings             0.5.1
websocket-client         1.2.3
wheel                    0.37.1
widgetsnbextension       3.5.1
win-inet-pton            1.1.0
wincertstore             0.2
wrapt                    1.13.3
xlrd                     2.0.1
yarl                     1.5.1
zipp                     3.7.0

(pytorch) C:\Users\Jayshow\PycharmProjects\spacy_transform>python -m spacy train textcatcorpus\config.cfg --output ./textcatoutput --paths.train
textcatcorpus\train.spacy --paths.dev textcatcorpus\dev.spacy --gpu-id 0
C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\numpy\.libs\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll
C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\numpy\.libs\libopenblas.GK7GX5KEQ4F6UYO3P26ULGBQYHGQO7J4.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"
ℹ Saving to output directory: textcatoutput
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================
[2022-02-12 20:50:40,286] [INFO] Set up nlp object from config
[2022-02-12 20:50:40,298] [INFO] Pipeline: ['textcat_multilabel']
[2022-02-12 20:50:40,304] [INFO] Created vocabulary
[2022-02-12 20:50:40,306] [INFO] Finished initializing nlp object
[2022-02-12 20:50:41,868] [INFO] Initialized pipeline components: ['textcat_multilabel']
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['textcat_multilabel']
ℹ Initial learn rate: 0.0001
E    #       LOSS TEXTC...  CATS_SCORE  SCORE
---  ------  -------------  ----------  ------
⚠ Aborting and saving the final best model. Encountered exception:
AttributeError("module 'cupy' has no attribute 'cublas'")
Traceback (most recent call last):
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\__main__.py", line 4, in <module>
    setup_cli()
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\cli\_util.py", line 71, in setup_cli
    command(prog_name=COMMAND)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\click\core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\click\core.py", line 782, in main
    rv = self.invoke(ctx)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\click\core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\click\core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\click\core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\typer\main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\cli\train.py", line 45, in train_cli
    train(config_path, output_path, use_gpu=use_gpu, overrides=overrides)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\cli\train.py", line 75, in train
    train_nlp(nlp, output_path, use_gpu=use_gpu, stdout=sys.stdout, stderr=sys.stderr)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\training\loop.py", line 122, in train
    raise e
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\training\loop.py", line 105, in train
    for batch, info, is_best_checkpoint in training_step_iterator:
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\spacy\training\loop.py", line 219, in train_while_improving
    proc.finish_update(optimizer)  # type: ignore[attr-defined]
  File "spacy\pipeline\trainable_pipe.pyx", line 252, in spacy.pipeline.trainable_pipe.TrainablePipe.finish_update
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\thinc\model.py", line 327, in finish_update
    param, grad = optimizer(
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\thinc\optimizers.py", line 230, in __call__
    gradient = ops.clip_gradient(gradient, self.grad_clip)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\thinc\backends\cupy_ops.py", line 107, in clip_gradient
    grad_norm = cupy.maximum(frobenius_norm(gradient), 1e-12)
  File "C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\thinc\backends\cupy_ops.py", line 105, in frobenius_norm
    return cupy.cublas.nrm2(X_vec)
AttributeError: module 'cupy' has no attribute 'cublas'

**Comment by polm** [comment]
Based on the errors and warnings here it seems like there's something wrong with your environment.

    WARNING: Ignoring invalid distribution -upy-cuda110 (c:\users\jayshow\anaconda3\envs\pytorch\lib\site-packages)

You should figure out where that's coming from and fix it.

```
C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\numpy_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\numpy.libs\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll
C:\Users\Jayshow\anaconda3\envs\pytorch\lib\site-packages\numpy.libs\libopenblas.GK7GX5KEQ4F6UYO3P26ULGBQYHGQO7J4.gfortran-win_amd64.dll
warnings.warn("loaded more than 1 DLL from .libs:"
```

This makes it look like something is wrong with your numpy or pytorch installs? 

Can you start over in a fresh environment and follow the instructions [here](https://spacy.io/usage)? You shouldn't have to try different versions of cuda-cupy - you should just get one that matches your cuda version and pip should install the latest compatible one.

**Comment by dariusbakunas** [comment, marked as answer]
This helped me: https://github.com/explosion/thinc/issues/920

`pip install thinc==8.2.3`

**Reply by Sturmharder** [reply]
Thanks!!! That helped me
